Wikipedia/Code/WMFArticleViewController.m

```diff
@@ -97,7 +97,11 @@ - (nullable UIActivityViewController *)sharingActivityViewControllerWithTextSnip
     NSURL *url = [NSURL wmf_desktopURLForURL:self.url];
 
     if (url) {
-        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", url.absoluteString, @"wprov=sfsi1"]]];
+        NSString *urlAsString = url.absoluteString;
+        if([urlAsString hasSuffix:@"!"]) {
+            urlAsString = [urlAsString stringByReplacingOccurrencesOfString:@"!" withString:@"%21"];
+        }
+        [items addObject:[[NSURL alloc] initWithString:[NSString stringWithFormat:@"%@?%@", urlAsString, @"wprov=sfsi1"]]];
```

Comment:
@nambatee the character set would need to be modified to remove the `!`. It'd look like this:
```
        NSMutableCharacterSet *characterSet = [[NSCharacterSet URLPathAllowedCharacterSet] mutableCopy];
        [characterSet removeCharactersInString:@"!"];
```